subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformIsolationTest.groovy

```diff
@@ -0,0 +1,194 @@
+                    println "variants: " + first.collect { it.variant.attributes }
+                }
+            }
+            
+            task increment {
+                doFirst {
+                    // Just to show that incrementing the counter doesn't matter.
+                    counter.increment()
+                }
+            }
+
+            task resolveSecond(type: Copy) {
+                def second = configurations.compile.incoming.artifactView {
+                    attributes { it.attribute(artifactType, 'secondCount') }
+                }.artifacts
+                from second.artifactFiles
+                into "\${buildDir}/libs2"
+                
+                doLast {
+                    println "files: " + second.collect { it.file.name }
+                    println "ids: " + second.collect { it.id }
+                    println "components: " + second.collect { it.id.componentIdentifier }
+                    println "variants: " + second.collect { it.variant.attributes }
+                }
+            }
+
+            task resolveThird(type: Copy) {
+                def third = configurations.compile.incoming.artifactView {
+                    attributes { it.attribute(artifactType, 'thirdCount') }
+                }.artifacts
+                from third.artifactFiles
+                into "\${buildDir}/libs3"
+                
+                doLast {
+                    println "files: " + third.collect { it.file.name }
+                    println "ids: " + third.collect { it.id }
+                    println "components: " + third.collect { it.id.componentIdentifier }
+                    println "variants: " + third.collect { it.variant.attributes }
+                }
+            }
+            
+            task resolve dependsOn 'resolveFirst', 'increment', 'resolveSecond', 'resolveThird'
+        """
+
+        when:
+        run 'resolve'
+
+        then:
+        outputContains("variants: [{artifactType=firstCount}, {artifactType=firstCount}]")
+        file("build/libs1").assertHasDescendants("test-1.3.jar.txt", "test2-2.3.jar.txt")
+        file("build/libs1/test-1.3.jar.txt").readLines() == ["0", "1", "2", "3", "4"]
+        file("build/libs1/test2-2.3.jar.txt").readLines() == ["0", "1", "2", "3", "4"]
+
+        and:
+        outputContains("variants: [{artifactType=secondCount}, {artifactType=secondCount}]")
+        file("build/libs2").assertHasDescendants("test-1.3.jar.txt", "test2-2.3.jar.txt")
+        file("build/libs2/test-1.3.jar.txt").readLines() == ["1", "2", "3", "4", "5"]
+        file("build/libs2/test2-2.3.jar.txt").readLines() == ["1", "2", "3", "4", "5"]
+
+        and:
+        outputContains("variants: [{artifactType=thirdCount}, {artifactType=thirdCount}]")
+        file("build/libs3").assertHasDescendants("test-1.3.jar.txt", "test2-2.3.jar.txt")
+        file("build/libs3/test-1.3.jar.txt").readLines() == ["2", "3", "4", "5", "6"]
```

Comment:
If I read the setup correctly, all these transforms should have exactly the same output. The arguments should be isolated the first time the transform is actually needed, not when it is created. Is this something we plan to address now or later?